libnetwork/drivers/bridge/internal/nftabler/endpoint.go

```diff
@@ -4,13 +4,63 @@ package nftabler
 
 import (
 	"context"
+	"fmt"
 	"net/netip"
+	"strings"
+
+	"github.com/docker/docker/libnetwork/drivers/bridge/internal/firewaller"
+	"github.com/docker/docker/libnetwork/internal/nftables"
 )
 
 func (n *network) AddEndpoint(ctx context.Context, epIPv4, epIPv6 netip.Addr) error {
-	return nil
+	return n.modEndpoint(ctx, epIPv4, epIPv6, true)
 }
 
 func (n *network) DelEndpoint(ctx context.Context, epIPv4, epIPv6 netip.Addr) error {
+	return n.modEndpoint(ctx, epIPv4, epIPv6, false)
+}
+
+func (n *network) modEndpoint(ctx context.Context, epIPv4, epIPv6 netip.Addr, enable bool) error {
+	if n.fw.config.IPv4 && epIPv4.IsValid() {
+		if err := n.filterDirectAccess(ctx, n.fw.table4, n.config.Config4, epIPv4, enable); err != nil {
+			return err
+		}
+		if err := nftApply(ctx, n.fw.table4); err != nil {
+			return fmt.Errorf("adding rules for bridge %s: %w", n.config.IfName, err)
+		}
+	}
+	if n.fw.config.IPv6 && epIPv6.IsValid() {
+		if err := n.filterDirectAccess(ctx, n.fw.table6, n.config.Config6, epIPv6, enable); err != nil {
+			return err
+		}
+		if err := nftApply(ctx, n.fw.table6); err != nil {
+			return fmt.Errorf("adding rules for bridge %s: %w", n.config.IfName, err)
+		}
+	}
 	return nil
 }
+
+// filterDirectAccess drops packets addressed directly to the container's IP address,
+// when direct routing is not permitted by network configuration.
+//
+// It is a no-op if:
+//   - gateway mode is "nat-unprotected" or "routed".
+//   - direct routing is enabled at the daemon level.
+//   - "raw" rules are disabled (possibly because the host doesn't have the necessary
+//     kernel support).
+//
+// Packets originating on the bridge's own interface and addressed directly to the
+// container are allowed - the host always has direct access to its own containers
+// (it doesn't need to use the port mapped to its own addresses, although it can).
+//
+// "Trusted interfaces" are treated in the same way as the bridge itself.
+func (n *network) filterDirectAccess(ctx context.Context, table nftables.TableRef, conf firewaller.NetworkConfigFam, epIP netip.Addr, enable bool) error {
+	if n.config.Internal || conf.Unprotected || conf.Routed || n.fw.config.AllowDirectRouting {
+		return nil
+	}
+	updater := table.ChainUpdateFunc(ctx, rawPreroutingChain, enable)
+	ifNames := strings.Join(append(n.config.TrustedHostInterfaces, n.config.IfName), ", ")
+	return updater(ctx, rawPreroutingPortsRuleGroup,
+		`%s daddr %s iifname != { %s } counter drop comment "DROP DIRECT ACCESS"`,
+		table.Family(), epIP, ifNames)
```

Comment:
While _possibly_ okay, it makes me nervous that the backing array of `n.config.TrustedHostInterfaces` may be mutated here. It's a surprising source of shared mutable state, one that is easily avoided.
```suggestion
	ifNames := strings.Join(n.config.TrustedHostInterfaces, ", ")
	return updater(ctx, rawPreroutingPortsRuleGroup,
		`%s daddr %s iifname != { %s, %s } counter drop comment "DROP DIRECT ACCESS"`,
		table.Family(), epIP, ifNames, n.config.IfName)
```

Reply:
Ok, I'm not entirely sure I see the issue, but nftables seems to tolerate the trailing comma when there are no trusted interfaces. So, done.

Reply:
`append(a, b)` mutates the backing array of `a` in-place if there is available capacity, which may lead to spooky action at a distance: [demonstration](https://go.dev/play/p/Bvwnl-O0MGm).

Reply:
> don't think it could have been a problem in this case

[It would still be a problem.](https://go.dev/play/p/EBEs106_OUL)

Reply:
> [It would still be a problem.](https://go.dev/play/p/EBEs106_OUL)

That would be a silly thing to do!

Reply:
> (There's always an ifname, but not normally any trusted interfaces. Your example shows the opposite.)

Ah, whoops! You are correct.